Add a module API to allow modules to add push rules for local users

babolivier opened the pull request:
This is happening in the context of the Tchap mainlining. See https://github.com/matrix-org/matrix-dinsic/issues/856 for more context.

clokep commented on synapse/module_api/__init__.py:
```diff
@@ -1340,6 +1349,91 @@ async def store_remote_3pid_association(
         """
         await self._store.add_user_bound_threepid(user_id, medium, address, id_server)
 
+    async def add_push_rule_for_user(
```
I think much of this method is copied from `PushRuleRestServlet` -- would it make sense to create a `PushRuleHandler` to encapsulate that logic and call it from both here and `PushRuleRestServlet`? (Parts of that class could probably be moved into the handler too...)

babolivier replied:
Yes probably.

babolivier commented:
Turns out this was the wrong approach, so I changed this PR to instead implement an API for modules to change the actions for a given rule. I've also moved a few things to a push rules handler, so the module API doesn't import directly from client resources.

richvdh commented on synapse/module_api/errors.py:
```diff
@@ -16,14 +16,18 @@
     "SynapseError",
     "ConfigError",
+    "InvalidRuleException",
+    "NotFoundError",
```
ugh, I'm not that enthusiastic about exposing NotFoundError. It's a bit of a poster-child for our antipattern of raising HTTP errors from the depths of database code.

I see we already have a `RuleNotFoundException`. Can we use that?

richvdh commented on synapse/handlers/push_rules.py:
```diff
@@ -0,0 +1,129 @@
+        self._notifier = hs.get_notifier()
+        self._store = hs.get_datastores().main
+
+    async def set_rule_attr(
+        self, user_id: str, spec: RuleSpec, val: Union[bool, JsonDict]
+    ) -> None:
+        """Set an attribute (enabled or actions) on an existing push rule.
+
+        Args:
+            user_id: the user for which to modify the push rule.
+            spec: the spec of the push rule to modify.
+            val: the value to change the attribute to.
+        """
```
it looks like you're relying on this raising particular exceptions in particular situations, so that's part of the calling contract for this method. Please could you details of the exceptions to the docstring?

richvdh commented on synapse/handlers/push_rules.py:
```diff
@@ -0,0 +1,129 @@
+    attr: Optional[str]
+
+
+class PushRulesHandler:
```
it'd be nice to give this a brief description in a docstring, even if it's just a oneliner.

richvdh commented on synapse/handlers/push_rules.py:
```diff
@@ -0,0 +1,129 @@
+        else:
+            raise UnrecognizedRequestError()
+
+    def notify_user(self, user_id: str) -> None:
```
docstring please

richvdh commented on synapse/rest/client/push_rule.py:
```diff
@@ -81,8 +78,12 @@ async def on_PUT(self, request: SynapseRequest, path: str) -> Tuple[int, JsonDic
+            except InvalidRuleException as e:
+                raise SynapseError(400, "Invalid actions: %s" % e)
+
+            self._push_rules_handler.notify_user(user_id)
```
shouldn't notify_user happen automatically as a side-effect of `set_rule_attr`?

richvdh commented on synapse/handlers/push_rules.py:
```diff
@@ -0,0 +1,129 @@
+            # a known attribute first.
+            raise UnrecognizedRequestError()
+
+        namespaced_rule_id = namespaced_rule_id_from_spec(spec)
```
I'm sorely tempted to inline this rather than lug `namespaced_rule_id_from_spec` around

```suggestion
        namespaced_rule_id = f"global/{spec.template}/{spec.rule_id}"
```

richvdh commented on synapse/server.py:
```diff
@@ -850,3 +851,7 @@ def get_request_ratelimiter(self) -> RequestRatelimiter:
             self.config.ratelimiting.rc_message,
             self.config.ratelimiting.rc_admin_redaction,
         )
+
+    @cache_in_self
+    def get_push_rules_handler(self) -> PushRulesHandler:
+        return PushRulesHandler(self)
```
put this next to the other Handlers, maybe?

richvdh commented on synapse/module_api/__init__.py:
```diff
@@ -1340,6 +1342,75 @@ async def store_remote_3pid_association(
+        """
+        try:
+            check_actions(actions)
+        except InvalidRuleException:
+            return False
```
given we're exposing `InvalidRuleException` anyway, why not let the exception propagate? it will give the caller more info than a simple boolean.

richvdh reviewed (changes requested):
seems sensible, but I have a few thoughts.

dklimpel commented:
Is there a plan to add something like this?
- "rewrite Sygnal URL"
   https://github.com/matrix-org/synapse-dinsic/pull/69

richvdh reviewed (commented):
looks broadly good. still a few nits.

richvdh commented on synapse/handlers/push_rules.py:
```diff
@@ -37,12 +37,13 @@ class PushRulesHandler:
         self, user_id: str, spec: RuleSpec, val: Union[bool, JsonDict]
     ) -> None:
-        """Set an attribute (enabled or actions) on an existing push rule.
+        """Set an attribute (enabled or actions) on an existing push rule, and notify
+        listeners (e.g. sync handler) of the change.
```
```suggestion
        """Set an attribute (enabled or actions) on an existing push rule.
        
        Notifies listeners (e.g. sync handler) of the change.
```

richvdh reviewed (approved):
lgtm!
